### PR TITLE
Enhance simple test

### LINKED ISSUE #3670: Enhance SimpleTest command.
**BUG REPORT**
The SimpleTest only test ledger write, didn't include read, and it didn't delete the test ledger.

### PR DESCRIPTION
**Pull request by horizonzy**
### Motivation
Fixes #3670

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java [hunk 1]
```diff
@@ -102,10 +101,12 @@ protected void run(BookKeeper bk, Flags flags) throws Exception {
                 }
             }
             LOG.info("{} entries written to ledger {}", flags.numEntries, wh.getId());
-            if (flags.cleanup) {
-                LOG.info("Cleaning up the ledger {}", wh.getId());
-                result(bk.newDeleteLedgerOp().withLedgerId(wh.getId()).execute());
+    
+            try (ReadHandle rh = result(bk.newOpenLedgerOp().withLedgerId(wh.getId()).withDigestType(DigestType.CRC32C)
+                    .withPassword(new byte[0]).execute())) {
+                rh.read(0, flags.numEntries);
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:107]: Would you please assert the result of rh.read()?

Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:107]: Fine, the write data is an empty array. I will change the written data and verify it.

**Comment by horizonzy** [comment]
rerun failure checks

**Review by hangc0276 — APPROVED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java [hunk 2]
```diff
@@ -61,8 +65,6 @@ public static class Flags extends CliFlags {
         private int ackQuorumSize = 2;
         @Parameter(names = { "-n", "--num-entries" }, description = "Entries to write (default 100)")
         private int numEntries = 100;
-        @Parameter(names = { "-c", "--clean-up" }, description = "Clean up ledger created after simple test")
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:64]: It is an easily way to generate a ledger. I am not sure if it should be deleted automatically. If we delete it automatically, the only way to get a ledger is writing a program, it looks very inconvenient.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java [hunk 3]
```diff
@@ -81,6 +83,7 @@ public SimpleTestCommand(Flags flags) {
     @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
     protected void run(BookKeeper bk, Flags flags) throws Exception {
         byte[] data = new byte[100]; // test data
+        Arrays.fill(data, (byte) '1');
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:86]: Fill them randomly?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java [hunk 4]
```diff
@@ -102,10 +105,15 @@ protected void run(BookKeeper bk, Flags flags) throws Exception {
...
-                result(bk.newDeleteLedgerOp().withLedgerId(wh.getId()).execute());
+
+            try (ReadHandle rh = result(bk.newOpenLedgerOp().withLedgerId(wh.getId()).withDigestType(DigestType.CRC32C)
+                    .withPassword(new byte[0]).execute())) {
+                LedgerEntries ledgerEntries = rh.read(0, flags.numEntries);
+                for (LedgerEntry ledgerEntry : ledgerEntries) {
+                    assert Arrays.equals(ledgerEntry.getEntryBytes(), data);
+                }
             }
+            result(bk.newDeleteLedgerOp().withLedgerId(wh.getId()).execute());
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:116]: In Finally?

**Review by zymap — CHANGES_REQUESTED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java (hunk 2 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:64]: This command is for testing the bookkeeper to create a ledger and read the ledger. If already cover the user's case.
If users want to test by themselves, they should create a new one and read it, not support the ledger generated by SimpleTestCommand.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java (hunk 3 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:86]: Fine

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java [hunk 5]
```diff
@@ -102,10 +105,15 @@ protected void run(BookKeeper bk, Flags flags) throws Exception {
...
             LOG.info("{} entries written to ledger {}", flags.numEntries, wh.getId());
-            if (flags.cleanup) {
-                LOG.info("Cleaning up the ledger {}", wh.getId());
-                result(bk.newDeleteLedgerOp().withLedgerId(wh.getId()).execute());
+
+            try (ReadHandle rh = result(bk.newOpenLedgerOp().withLedgerId(wh.getId()).withDigestType(DigestType.CRC32C)
+                    .withPassword(new byte[0]).execute())) {
+                LedgerEntries ledgerEntries = rh.read(0, flags.numEntries);
+                for (LedgerEntry ledgerEntry : ledgerEntries) {
+                    assert Arrays.equals(ledgerEntry.getEntryBytes(), data);
```
[the remark this answers is not in the document]
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:113]: Good point.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java (hunk 4 above)
Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/client/SimpleTestCommand.java:116]: Fine.

**Review by dlg99 — APPROVED** [review]

**Review by zymap — APPROVED** [review]

**Comment by lhotari** [comment]
I'm fixing the integration tests in #4197. It seems that simpletest will always fail and has never been working.

There are at least 2 problems:
- the last entry id parameter to the read method is invalid (off by one)
- the LAC is updated asynchronously while the ledger is open. The read could fail if the entries aren't yet confirmed.

Fixing in https://github.com/lhotari/bookkeeper/commit/6ffda5cb since I cannot get integration tests to pass without it.
This is a sign that integration tests have been broken for a very long time on apache/bookkeeper.

. The request to read the entries is

**Pull request merged**